sdks/python/apache_beam/dataframe/io.py

```diff
@@ -0,0 +1,178 @@
+
+import apache_beam as beam
+from apache_beam import io
+from apache_beam.dataframe import frame_base
+
+
+def read_csv(path, *args, **kwargs):
+  return _ReadFromPandas(pd.read_csv, path, args, kwargs)
+
+
+def write_csv(df, path, *args, **kwargs):
+  from apache_beam.dataframe import convert
+  # TODO(roberwb): Amortize the computation for multiple writes?
+  return convert.to_pcollection(df) | _WriteToPandas(
+      pd.DataFrame.to_csv, path, args, kwargs, incremental=True, binary=False)
+
+
+def _prefix_range_index_with(prefix, df):
+  if isinstance(df.index, pd.RangeIndex):
+    return df.set_index(prefix + df.index.map(str).astype(str))
+  else:
+    return df
+
+
+class _ReadFromPandas(beam.PTransform):
+  def __init__(self, reader, path, args, kwargs):
+    if not isinstance(path, str):
+      raise frame_base.WontImplementError('non-deferred')
+    self.reader = reader
+    self.path = path
+    self.args = args
+    self.kwargs = kwargs
+
+  def expand(self, root):
+    # TODO(robertwb): Handle streaming (with explicit schema).
+    paths_pcoll = root | beam.Create([self.path])
+    first = io.filesystems.FileSystems.match([self.path],
+                                             limits=[1
+                                                     ])[0].metadata_list[0].path
+    with io.filesystems.FileSystems.open(first) as handle:
+      df = next(self.reader(handle, *self.args, chunksize=100, **self.kwargs))
+
+    # TODO(robertwb): Actually make an SDF.
+    def expand_pattern(pattern):
+      for match_result in io.filesystems.FileSystems.match([pattern]):
+        for metadata in match_result.metadata_list:
+          yield metadata.path
+
+    pcoll = (
+        paths_pcoll
+        | beam.FlatMap(expand_pattern)
+        | beam.ParDo(_ReadFromPandasDoFn(self.reader, self.args, self.kwargs)))
+    from apache_beam.dataframe import convert
+    return convert.to_dataframe(
+        pcoll, proxy=_prefix_range_index_with(':', df[:0]))
+
+
+class _ReadFromPandasDoFn(beam.DoFn):
+  def __init__(self, reader, args, kwargs):
+    # avoid pickling issues
+    self.reader = reader.__name__
+    self.args = args
+    self.kwargs = kwargs
+
+  def process(self, path):
+    reader = getattr(pd, self.reader)
+    for df in reader(path, *self.args, chunksize=100, **self.kwargs):
+      yield _prefix_range_index_with(path + ':', df)
+
+
+class _WriteToPandas(beam.PTransform):
+  def __init__(
+      self, writer, path, args, kwargs, incremental=False, binary=True):
+    self.writer = writer
+    self.path = path
+    self.args = args
+    self.kwargs = kwargs
+    self.incremental = incremental
+    self.binary = binary
+
+  def expand(self, pcoll):
+    return pcoll | io.Write(
+        _WriteToPandasFileBasedSink(
+            self.writer,
+            self.path,
+            self.args,
+            self.kwargs,
+            self.incremental,
+            self.binary))
+
+
+class _WriteToPandasFileBasedSink(io.FileBasedSink):
```

Comment:
Instead of extending FileBasedSink here please implement a fileio.FileSink and use fileio.WriteToFiles.
https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/fileio.py#L76

Reply:
Done.